lib/connection.js

```diff
@@ -168,15 +163,33 @@ Connection.prototype.receiveResponse = function(data) {
 /**
  * Close this connection
  */
-Connection.prototype.close = function() {
+Connection.prototype.close = function(error) {
 
-    if (this.stream) {
+    if (!this.connected) {
+        return;
+    }
+
+    if (this.stream && this.stream.writable) {
         this.stream.end();
     }
+
     this.connected = false;
-    this.disconnected = true;
     this.stream = null;
     this.emit('close');
+
+    var hasRequests = this.requestBuffer.length;
+    if (hasRequests || error) {
+        this.emit('error.server', error || 'Server unexpectedly closed network connection.');
```

Comment:
I think this should be `error.network` so it can be distinguished from other server errors if needed.